LanguageFeatures/Patterns/type_inference_A32_t05.dart

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// If p with required type T is in an irrefutable context:
+/// - It is a compile-time error if M is not assignable to T.
+/// - Else if M is not a subtype of T then an implicit coercion or cast is
+///   inserted before the pattern binds the value, tests the value's type,
+///   destructures the value, or invokes a function with the value as a target
+///   or argument.
+///
+/// @description Check that it is a compile-time error if `M` is not assignable
+/// to `T` in irrefutable context. Test object patterns
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
```

Comment:
The comments in LanguageFeatures/Patterns/type_inference_A32_t02.dart should be applicable here as well.

Reply:
The matched value type in line 24 is `Shape<T>` for a `T` which is obtained by type inference.

Given that we're not testing phase 1 (computation of the pattern type schema) and phase 2 (inference on the initializing expression), it's better if the initializing expression has a firm type: All type arguments should be specified in the initializing expression.

(We should already have other tests testing phase 1 and 2.)

For the pattern, when testing "matched value type isn't assignable", we should also have a firm type, because then there's no doubt at all about (1) the matched value type and the pattern required type can be seen in the source code directly, and (2) from that we can easily see whether the assignability holds or not, so we can easily choose the correct expectation.

With object patterns we can have nested cases as well as top-level cases.

```dart
var Square(...) = Shape(); // OK, no type arguments after all, this must fail no matter what. ;-)
var Square<T1>(...) = Square<T2>(); // Top-level case, assignability determined by `T1`/`T2`.
var Square<T3>(area: T5 x1) = throw 0; // `area` has some return type, let's call it `T4`.
```

The third case gives rise to a matched value type from the return type `T4` of the getter `area`, which might again depend on the value of `T3`. We must then test that `T5` causes an error unless `T4` is assignable to `T5`. The initializing expression has type `Never` because it does not matter (and we just want it to be type correct and not interfere with the relationship between `T4` and `T5`).

This is again the shortest possible test: From `T3` to `T5`, we avoid inferring `T3` from the initializing expression, and we avoid inferring `T5` from the return type of `area` (which wouldn't allow us to test that the compile-time error is actually reported, so that's another reason why we don't allow `x1` to have an inferred type).

Reply:
The main thing here is that the nested case is missing.